pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -53,39 +77,71 @@ func NewNode(constraints *v1alpha5.Constraints, instanceTypeOptions []cloudprovi
+		return err
 	}
 
-	podRequirements := v1alpha5.NewPodRequirements(pod)
-	if err := n.Constraints.Requirements.Compatible(podRequirements); err != nil {
+	// Taints should only arrive via the provisioner, so this should never occur but to be safe
+	if err := n.Taints.Tolerates(pod); err != nil {
 		return err
 	}
 
-	tightened := n.Constraints.Requirements.Add(podRequirements.Requirements...)
+	tightened := n.Requirements.Add(podRequirements.Requirements...)
 	// Ensure that at least one instance type of the instance types that we are already narrowed down to based on the
 	// existing pods can support the pod resources and combined pod + provider requirements
+	podRequests := resources.RequestsForPods(pod)
 	for _, it := range n.InstanceTypeOptions {
-		if cloudprovider.Compatible(it, tightened) && n.hasCompatibleResources(resources.RequestsForPods(pod), it) {
+		newSize := n.reservedResources(it, podRequests)
+		if cloudprovider.Compatible(it, tightened) &&
+			n.newPodCanFit(newSize, it) &&
+			n.hasCompatibleResources(podRequests, it) {
 			return nil
 		}
 	}
 	return errors.New("no matching instance type found")
 }
 
+func (n Node) reservedResources(it cloudprovider.InstanceType, extra v1.ResourceList) v1.ResourceList {
+	return resources.Merge(it.Overhead(), n.daemonResources, n.podResources, extra)
+}
+
+func (n *Node) newPodCanFit(newSize v1.ResourceList, it cloudprovider.InstanceType) bool {
+	for resourceName, totalQuantity := range it.Resources() {
+		reservedQuantity := newSize[resourceName]
+		if reservedQuantity.Cmp(totalQuantity) > 0 {
+			return false
+		}
+	}
+
+	instancePodMax := it.Resources()[v1.ResourcePods]
+	if !instancePodMax.IsZero() && instancePodMax.CmpInt64(int64(len(n.Pods)+1)) < 0 {
+		return false
+	}
+	return true
+}
+
 // Add adds a pod to the Node which tightens constraints, possibly reducing the available instance type options for this
 // node
 func (n *Node) Add(pod *v1.Pod) {
-	n.Pods = append(n.Pods, pod)
-	n.Constraints = n.Constraints.Tighten(pod)
+	n.Requirements.Combine(v1alpha5.NewPodRequirements(pod).Requirements...)
+
+	podRequests := resources.RequestsForPods(pod)
 	var instanceTypeOptions []cloudprovider.InstanceType
 	for _, it := range n.InstanceTypeOptions {
-		if cloudprovider.Compatible(it, n.Constraints.Requirements) &&
-			n.hasCompatibleResources(resources.RequestsForPods(pod), it) {
+		newSize := n.reservedResources(it, podRequests)
+		if cloudprovider.Compatible(it, n.Requirements) &&
+			n.newPodCanFit(newSize, it) &&
+			n.hasCompatibleResources(podRequests, it) {
 			instanceTypeOptions = append(instanceTypeOptions, it)
 		}
 	}
+	// have to add the pod after filtering instance types as newPodCanFit() checks if a new pod can fit, including the
+	// pod count
+	n.Pods = append(n.Pods, pod)
 	n.InstanceTypeOptions = instanceTypeOptions
+	n.podResources = resources.Merge(n.podResources, resources.RequestsForPods(pod))
+	n.recalculateDaemonResources()
```

Comment:
Why do we need to recalculate daemon resources after every pod added? Is this only for the case that another daemonset is created during the bin-packing logic?

Reply:
Good question, it's to handle the case of a pod making a daemonset unschedulable. You can have a daemonset that runs on arm only, or only runs in test-zone-2.  If we add a pod that has an amd64, or runs in test-zone-1, that daemonset will no longer schedule to the pod and we can use the resources it would use for more pods.